Add "Lightbend Orchestration for Kubernetes" to deploy to Kubernetes

longshorej opened the pull request:
This removes the ConductR plugin and replaces it with [sbt-reactive-app](https://github.com/lightbend/sbt-reactive-app). It also adds a section to the readme showing how this tooling can be used to manually generate and apply Kubernetes YAML (Typical production env), or how it can be automated for local development via `sbt "deploy minikube"`.

I suspect we may have to discuss further out of band, but wanted to get this PR up for now.

ignasi35 commented:
This README is getting out of hand, we should (in a separate PR) split it and/or add a TOC and /or ...

ignasi35 commented:
Just noticed: 

```
[warn] 	* com.typesafe.akka:akka-stream_2.12:2.5.9 is selected over {2.5.7, 2.4.20, 2.5.3, 2.5.8}
[warn] 	    +- com.typesafe.akka:akka-persistence-query_2.12:2.5.9 () (depends on 2.5.9)
[warn] 	    +- com.lightbend.lagom:lagom-logback_2.12:1.4.0       (depends on 2.5.9)
```

in the build stdout. We'll need to track down that dependency to akka-stream `2.4.20`.

ignasi35 commented on README.md:
````diff
@@ -54,3 +54,190 @@ For a simple, gentler, introduction to Lagom, have a look at the [Hello World gi
+
+```bash
+echo "http://$(minikube ip)"
+```
````
same as above.

ignasi35 commented:
After talking with @longshorej we noticed the error I had was due to Ingress not being ready. The following worked on my laptop:

```
  minikube version
  rp version
  minikube start --memory 6000
  minikube addons enable ingress   # this is missing in the docs.
  sbt "deploy minikube"
  open "http://$(minikube ip)"
```

longmuir commented on README.md:
````diff
@@ -54,3 +54,190 @@ For a simple, gentler, introduction to Lagom, have a look at the [Hello World gi
+
+```bash
+# Secret values, be sure to change them
+
````
Add a link or even the preferred command to generate these values. Looks like this is the page for Play, I'm not sure if there's more relevant docs: https://www.playframework.com/documentation/latest/ApplicationSecret

longmuir commented on README.md:
```diff
@@ -54,3 +54,190 @@ For a simple, gentler, introduction to Lagom, have a look at the [Hello World gi
+export service_elasticsearch=_http._tcp.reactive-sandbox-elasticsearch.default.svc.cluster.local
+
+# Configure the services to allow requests to the minikube IP (Play's Allowed Hosts Filter)
+
```
Add a link to the Play doc for this as well. This I believe: https://www.playframework.com/documentation/latest/AllowedHostsFilter

longmuir commented on README.md:
````diff
@@ -54,3 +54,190 @@ For a simple, gentler, introduction to Lagom, have a look at the [Hello World gi
+
+```bash
+eval $(minikube docker-env)
+```
````
Could add a link "For more information..." that points to https://kubernetes.io/docs/getting-started-guides/minikube/#reusing-the-docker-daemon.

longmuir commented on README.md:
````diff
@@ -54,3 +54,190 @@ For a simple, gentler, introduction to Lagom, have a look at the [Hello World gi
+##### Build Project
+
+```bash
+sbt clean docker:publishLocal
````
This will build all of the applications in the project. It'd be good to also provide the command(s) for building the individual applications in the same manner that there are individual deployment commands below.

longmuir commented:
Agree the readme is getting large, but I wouldn't skimp on details. This is quite a complex process for those not that familiar with all of the pieces. A TOC might help. You could also perhaps split it between a more developer focused and a more ops focused guide. We might even consider making this into a guide on the tech hub.

ignasi35 commented:
Quick note: on my tests I only checked the `Deployment Workflow` not the `Development Workflow`

TimMoore commented:
Agree with splitting up into multiple documents. For comparison, [`online-auction-java`](https://github.com/lagom/online-auction-java) has [a section in its README about running in ConductR](https://github.com/lagom/online-auction-java#running-in-production-mode) that links off to [a separate document with detailed instructions](https://github.com/lagom/online-auction-java/blob/master/docs/running-in-conductr.md).

@rstento wrote a nice [Readme Template](https://docs.google.com/document/d/1-sbFgQi1nTObMppTFSju_06AIILJtKub-BBS2dEtWWg/edit?usp=sharing) document (only accessible to Lightbend employees) that is worth considering for this. It also suggests a table of contents, and some other formatting guidelines to help people orient themselves. I requested her review on this pull request as well.

TimMoore commented:
I've done some more digging into the ingress problems I mentioned above, and I have a better understanding of what's happening now.

The problem is the `--ingress-path-suffix '*'` flag specified in the guide. When you have that, it creates these nginx directives:

```
        location /* {

            # ... trimmed
            proxy_pass http://default-webgateway-10000;

        }

        location / {

            # ... trimmed
            proxy_pass http://upstream-default-backend;

        }
```

So there is `/` without a wildcard, and `/*` with it. It looks like nginx does prefix matching by default and _does not_ expect a `*` in the location. I believe it will treat that as a literal asterisk.

Running the command again, but without the suffix flag, works as expected:

```bash
rp generate-kubernetes-resources \
  --generate-ingress --ingress-name online-auction \
  webgateway:1.0.0-SNAPSHOT \
  searchimpl:1.0.0-SNAPSHOT \
  userimpl:1.0.0-SNAPSHOT \
  itemimpl:1.0.0-SNAPSHOT \
  biddingimpl:1.0.0-SNAPSHOT \
  webgateway:1.0.0-SNAPSHOT | kubectl apply -f -
```

So I wondered why I didn't have any problems when running `sbt "deploy minikube"` yesterday. I deleted and recreated everything that way, and discovered that this creates two different ingress resources, but without the trailing wildcard:

```
$ kubectl get ingress
NAME         HOSTS     ADDRESS          PORTS     AGE
search       *         192.168.99.100   80        1h
webgateway   *         192.168.99.100   80        1h

$ kubectl describe ingress search
Name:             search
Namespace:        default
Address:          192.168.99.100
Default backend:  default-http-backend:80 (172.17.0.3:8080)
Rules:
  Host  Path  Backends
  ----  ----  --------
  *     
        /api/search   search:10000 (<none>)
Annotations:
Events:
  Type    Reason  Age   From                      Message
  ----    ------  ----  ----                      -------
  Normal  CREATE  1h    nginx-ingress-controller  Ingress default/search
  Normal  UPDATE  1h    nginx-ingress-controller  Ingress default/search

$ kubectl describe ingress webgateway
Name:             webgateway
Namespace:        default
Address:          192.168.99.100
Default backend:  default-http-backend:80 (172.17.0.3:8080)
Rules:
  Host  Path  Backends
  ----  ----  --------
  *     
        /   webgateway:10000 (<none>)
Annotations:
Events:
  Type    Reason  Age   From                      Message
  ----    ------  ----  ----                      -------
  Normal  CREATE  1h    nginx-ingress-controller  Ingress default/webgateway
  Normal  UPDATE  1h    nginx-ingress-controller  Ingress default/webgateway
```

It surprised me that this works: it looks like they conflict, but it seems to be able to merge these into a coherent nginx configuration. The generated config is the same as the one generated by the command above, other than the ingress name variables and minor differences such as IP address. Both the web app and the search endpoint respond as expected.

Note: the search functionality is not yet implemented in the UI, but you can test it by going to http://$(minikube ip)/api/search?pageNo=0&pageSize=10 in your browser and you should get a `DeserializationException`. The exception is expected because this call expects a request body, but it proves that it is routing the request to the correct backend service.

I copied the two different generated nginx configurations and a diff between them to https://gist.github.com/TimMoore/1f443dfe723128d4c78fd2745a7f4728

TimMoore commented on README.md:
```diff
@@ -54,3 +54,190 @@ For a simple, gentler, introduction to Lagom, have a look at the [Hello World gi
+# Deploy ingress for everything
+
+rp generate-kubernetes-resources \
+  --ingress-path-suffix '*' \
```
As mentioned in my long comment on the main thread, the `--ingress-path-suffix '*'` flag should be removed.

renatocaval commented on README.md:
```diff
@@ -54,3 +54,190 @@ For a simple, gentler, introduction to Lagom, have a look at the [Hello World gi
+* [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl)
+* [Minikube](https://github.com/kubernetes/minikube) v0.25.0 or later (Verify with `minikube version`)
+* [Helm](https://github.com/kubernetes/helm)
+* [reactive-cli](https://github.com/lightbend/reactive-cli) 0.9.0 or later (Verify with `rp version`)
```
For the reactive-cli setup, it's better to point people to https://developer.lightbend.com/docs/reactive-platform-tooling/latest/cli-installation.html#install-the-cli instead. 

The github page has information about how to develop it.

renatocaval commented on README.md:
```diff
@@ -54,3 +54,190 @@ For a simple, gentler, introduction to Lagom, have a look at the [Hello World gi
+
+`minikube start --memory 6000`
+
+`sbt "deploy minikube"`
```
I'm getting this error:

```
java.lang.RuntimeException: reactive-cli is not installed [132]
	at scala.sys.package$.error(package.scala:27)
	at com.lightbend.rp.sbtreactiveapp.cmd.package$.runSuccess(package.scala:63)
	at com.lightbend.rp.sbtreactiveapp.cmd.rp$.assert(rp.scala:24)
	at com.lightbend.rp.sbtreactiveapp.BasicApp$$anonfun$projectSettings$37$$anonfun$apply$15$$anonfun$apply$16.apply(App.scala:260)
	at com.lightbend.rp.sbtreactiveapp.BasicApp$$anonfun$projectSettings$37$$anonfun$apply$15$$anonfun$apply$16.apply(App.scala:231)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

I have `reactive-cli` installed, but the symlink created by homebrew is `rp` instead of `reactive-cli`.

longshorej replied:
> It succeeded the second time. It's possible that the first time overlapped with a network glitch that I had, but since it's all running locally I wouldn't have expected it to affect anything. In any case, it's hard to tell from the output what the actual problem is or how to troubleshoot it.

It has to pull down reactive sandbox during that step from Bintray, which is about 200MB, so if there was a network glitch that would indeed cause the problem. 

I think we could at least guide the user to kubectl in that event, I've filed https://github.com/lightbend/sbt-reactive-app/issues/91 to track that.

--------------------------------

> When I run rp version, I get: [1] 37496 illegal hardware instruction rp version.

Hrm - I'm curious what CPU/mac model and macOS version?

longshorej commented:
Thanks everyone for the extensive feedback. I'll be working on addressing it today.

@renatocaval should/could we meet tomorrow to debug that?

rstento commented:
@longshorej, I'll wait until you've addressed comments and then review, if
you'd like?

On Tue, Feb 20, 2018 at 11:30 AM Jason Longshore <notifications@github.com>
wrote:

> Thanks everyone for the extensive feedback. I'll be working on addressing
> it today.
>
> @renatocaval <https://github.com/renatocaval> should/could we meet
> tomorrow to debug that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lagom/online-auction-scala/pull/62#issuecomment-367054419>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV1De49jifn_f1HclgUfI5BOc-Plf5inks5tWwE8gaJpZM4SC2HA>
> .
>

longshorej commented:
Yes, sounds good @rstento, I'll ping you when ready. Thanks!

longshorej commented on README.md:
````diff
@@ -54,3 +54,190 @@ For a simple, gentler, introduction to Lagom, have a look at the [Hello World gi
+Finally, you're ready to deploy the services. Be sure to adjust the secret variables and cassanda service address as necessary.
+
+```bash
+# Secret values, be sure to change them
````
I had a customer call recently where their deployment was failing because they didn't specify the application secret. This provided a good opportunity to explain the important of this value. I am in agreement with failing fast. I'll change these to `changeme` to support that.

I agree it can be tricky to figure out why it's failing, but I think it's a good learning experience if you're going to be using K8s.

longshorej commented:
I've addressed the feedback (Thanks everyone for the great feedback!). I've also split the Kubernetes stuff into `KUBERNETES.md` and linked to it from `README.md`.

@rstento this is ready for a review when you have some time. Thanks!

rstento commented on KUBERNETES.md:
```diff
@@ -0,0 +1,231 @@
+## Running: Kubernetes
```
@longshorej, it's not clear to me how this readme will be linked to the other one. It would be nice to have the intro of the primary readme explain that there are instructions for deploying the sample to Kubernetes and then link to this?

longshorej replied:
Yeah - there's a link from `README.md` to `KUBERNETES.md`, see here: https://github.com/longshorej/online-auction-scala-orchestration-example/blob/8f40fd4064d8e8797a27ba983fb575c70ea9ecd7/README.md

rstento commented on KUBERNETES.md:
```diff
@@ -0,0 +1,231 @@
+## Running: Kubernetes
+
+This project uses [Lightbend's Platform Tooling](https://developer.lightbend.com/docs/reactive-platform-tooling/latest/) to
+simplify deployment to [Kubernetes](https://kubernetes.io/). Follow the steps below to install Online Auction in your
```
"simplify deployment of the Lagom online auction to ....

longshorej replied:
done

rstento commented on KUBERNETES.md:
```diff
@@ -0,0 +1,231 @@
+* [Helm](https://github.com/kubernetes/helm)
+* [reactive-cli](https://developer.lightbend.com/docs/reactive-platform-tooling/latest/cli-installation.html#install-the-cli) 0.9.0 or later (Verify with `rp version`)
+
+## Deploy
```
I suggest changing this first heading from "Deploy" to "Choose deployment type" Then, maybe change the introduction to:
Once you've set up your environment, you can deploy online auction to Kubernetes. The steps are differ depending on whether you want to deploy in a development or production environment.

During development, it is simpler to deploy all services with a single sbt task. The <link>Developer Workflow section describes how to do this. It will build Docker images .... (include the description moved from the section into this intro).

The <link>Operations Workflow section describes the steps for deploying in a production environment. You will use sbt to build the images and the reactive-cli generates ... (pull up explanation from below). Because the production environment is more complex, additional steps are required as described below.

longshorej replied:
incorporated

rstento commented on KUBERNETES.md:
```diff
@@ -0,0 +1,231 @@
+
+> Want a fresh Minikube? Run `minikube delete` before the steps below.
+
+##### Start Minikube
```
I would start numbering the rest of the headings from this point on for the Operations Workflow. i.e. #### 1) Start Minikube,  And, in the intro, explain that it is a N-step process.

#### 2) Set up Docker engine ....

longshorej replied:
:+1:

rstento commented on KUBERNETES.md:
```diff
@@ -0,0 +1,231 @@
+
+##### Install Reactive Sandbox (optional; skip if you have your own Cassandra, Kafka, and Elasticsearch clusters) 
+
+The `reactive-sandbox` includes development-grade (i.e. it will lose your data) installations of Cassandra, 
```
Maybe, "The `reactive-sandbox` allows you to easily test your deployment with Cassandra, Kafka, Zookeeper, and Elasticsearch without using production versions.  Because it is a sandbox, it does not persist data across restarts. The sandbox is packaged as a Helm chart for easy installation into your Kubernetes cluster."

longshorej replied:
incorporated

longshorej commented on KUBERNETES.md:
```diff
@@ -0,0 +1,231 @@
+To build a single subproject, prefix the `docker:publishLocal` task with the project's name and a forward slash. For instance,
+the following will only build `itemImpl`:
+
+``bash
```
good catch :+1:

longshorej commented on KUBERNETES.md:
````diff
@@ -0,0 +1,231 @@
+Verify that Helm is available (this takes a minute or two):
+
+```bash
+kubectl --namespace kube-system get deploy/tiller-deploy
````
done

longshorej commented on KUBERNETES.md:
```diff
@@ -0,0 +1,231 @@
+# NOTE: You must change the secret values below or the applications will crash.
+#
+# These values are used for Play's application secret. It is important that they are set to a secret value.
+# More information: https://www.playframework.com/documentation/latest/ApplicationSecret
```
A random (!) thought I had was: Maybe Play could specialcase `random`? Similar to `changeme`, but instead of failing, replacing it with a random UUID or something? Or alternatively, our guides could pass some random data in as part of the bash scripts.

Leaving it as-is for now, as it's probably not really worth the effort to change.

TimMoore replied:
The thing is, you really need to make sure that this stays consistent between executions, or else it won't be able to decrypt any cookies.

This is really only an issue for Play webapps, and Lagom services ideally shouldn't need to set a secret at all unless you invoke functionality that requires it (rare for Lagom services). See https://github.com/playframework/playframework/issues/8075

In a perfect world, we'd demonstrate how to provision this using k8s secrets (or something more secure).

ignasi35 commented:
I made a couple of mistakes and knowing commands like `kubectl get all`, `kubectl logs <podname>` or `kubectl get ing`  but this is not a `kubectl` or a `minikube` tutorial/reference. 

Thanks @longshorej for the timely support!

longshorej commented:
:fireworks:
